src/Engine/Network/UDPClient.cpp

```diff
@@ -72,6 +84,72 @@ int UDPClient::readBuffer()
     return bytesReceived;
 }
 
+void UDPClient::readPartOfPacket()
+{
+    if (!m_Socket) {
+        return;
+    }
+    boost::system::error_code error;
+    // Peek header
+    m_Socket->receive(boost
+        ::asio::buffer((void*)m_ReadBuffer, 5 * sizeof(int)),
+        boost::asio::ip::udp::socket::message_peek, error);
+
+    int sizeOfPacket = 0;
+    memcpy(&sizeOfPacket, m_ReadBuffer, sizeof(int));
+    if (sizeOfPacket == 0) {
+        return;
+    }
+    int packetGroup = 0;
+    memcpy(&packetGroup, m_ReadBuffer + sizeof(int), sizeof(int));
+    int packetGroupIndex = 0;
+    memcpy(&packetGroupIndex, m_ReadBuffer + 2 * sizeof(int), sizeof(int));
+    int packetGroupSize = 0;
+    memcpy(&packetGroupSize, m_ReadBuffer + 3 * sizeof(int), sizeof(int));
+    //LOG_INFO("Packet group: %i. Group index: %i. Group size: %i. Packet size: %i.", packetGroup, packetGroupIndex, packetGroupSize, sizeOfPacket);
+    if (sizeOfPacket > m_Socket->available()) {
+        LOG_WARNING("UDPClient::readBuffer(): We haven't got the whole packet yet.");
+        // return;
+    }
+    // if the buffer is to small increase the size of it
+    boost::shared_ptr<char> packetData(new char[sizeOfPacket]);
+
+    // Read the message
+    size_t bytesReceived = m_Socket->receive_from(boost
+        ::asio::buffer((void*)(packetData.get()),
+            sizeOfPacket),
+        m_ReceiverEndpoint, 0, error);
+    if (error) {
+        LOG_ERROR("UDPClient::readPartOfPacket: %s", error.message().c_str());
+    }
+    // Might want to do this earlier when i figure out a good way to 
+    // remove data from network buffer.
+    if (hasReceivedPacket(packetGroup, packetGroupIndex)) {
+        return;
+    }
+    //std::map<unsigned int, std::vector<std::pair<unsigned int, boost::shared_ptr<char>>>> packetSegmentMap;
+    // If group exists
+    if (m_PacketSegmentMap.find(packetGroup) != m_PacketSegmentMap.end()) {
+        m_PacketSegmentMap.at(packetGroup).push_back(std::make_pair(packetGroupIndex, std::move(packetData)));
+    } else { // Create group and add element
+        m_PacketSegmentMap[packetGroup].push_back(std::make_pair(packetGroupIndex, std::move(packetData)));
+    }
+    return;
+}
+
+bool UDPClient::hasReceivedPacket(int packetGroup, int groupIndex)
+{
+    if (m_PacketSegmentMap.find(packetGroup) != m_PacketSegmentMap.end()) {
+        const std::vector<std::pair<int, boost::shared_ptr<char>>>& loopPacketGroup = m_PacketSegmentMap.at(packetGroup);
```

Comment:
Same thing here we can probably `.find()` once and reuse the iterator instead of doing `.at()`.